hardware/spectrometer/oceanoptics_spectrometer.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+"""
+This module controls spectrometers from Ocean Optics Inc.
+All spectrometers supported by python-seabreeze should work.
+
+Do "conda install -c poehlmann python-seabreeze"
+before using this module.
+
+Qudi is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Qudi. If not, see <http://www.gnu.org/licenses/>.
+
+Derived from the pyPL project (https://github.com/kaseyrussell/pyPL)
+Copyright 2010 Kasey Russell ( email: krussell _at_ post.harvard.edu )
+
+Copyright (c) the Qudi Developers. See the COPYRIGHT.txt file at the
+top-level directory of this distribution and at <https://github.com/Ulm-IQO/qudi/>
+
+"""
+
+from core.module import Base, ConfigOption
+from interface.spectrometer_interface import SpectrometerInterface
+import numpy as np
+import seabreeze.spectrometers as sb
+
+
+
+class OceanOptics(Base, SpectrometerInterface):
+    """ Hardware module for reading spectra from the WinSpec32 spectrometer software.
+
+    Example config for copy-paste:
+
+    myspectrometer:
+        module.Class: 'spectrometer.oceanoptics_spectrometer.OceanOptics'
+        spectrometer_serial: 'QEP01583' #insert here the right serial number.
+
+    """
+    _serial = ConfigOption('spectrometer_serial', missing='warn')
+
+    def on_activate(self):
+        """ Activate module.
+        """
+        self.integration_time = 1e4
+
+        self.spec = sb.Spectrometer.from_serial_number(self._serial)
+        self.log.info(''.format(self.spec.model, self.spec.serial_number))
+        self.spec.integration_time_micros(self.integration_time)
+
+    def on_deactivate(self):
+        """ Deactivate module.
+        """
+        self.spec.close()
+
+    def recordSpectrum(self):
+        """ Record spectrum from WinSpec32 software.
+
+            @return []: spectrum data
+        """
+        wavelengths = self.spec.wavelengths()
+        specdata = np.empty((2, len(wavelengths)), dtype=np.double)
+        specdata[0] = wavelengths/1e9
+        specdata[1] = self.spec.intensities()
+        return specdata
+
+    def getExposure(self):
+        """ Get exposure.
+
+            @return float: exposure
+
+            Not implemented.
+        """
+        return self.integration_time_micros
```

Comment:
Right after `__init__` this module instance has no attribute `integration_time_micros` and thus will raise `AttributeError`.
Always declare all instance variables in `__init__` first to avoid such a thing.

But I don't see why this variable is necessary? `integration_time_micros` seems to be stored in `spec` as well. So why not simply set and get this value in/from `spec` and avoid an additional copy which can possibly deviate from the true value set in the device?

Reply:
fixed with StatusVar